File: alchemist/alchemist-incarnation-biochemistry/src/main/java/it/unibo/alchemist/model/interfaces/CircularDeformableCell.java
/*
 * Copyright (C) 2010-2019, Danilo Pianini and contributors listed in the main project's alchemist/build.gradle file.
 *
 * This file is part of Alchemist, and is distributed under the terms of the
 * GNU General Public License, with a linking exception,
 * as described in the file LICENSE in the Alchemist distribution's top directory.
 */
package it.unibo.alchemist.model.interfaces;

/**
 * Implements a circular deformable cell.
 *
 * @param <P>
 *
 */
public interface CircularDeformableCell<P extends Position<? extends P>> extends CellWithCircularArea<P> {

    /**
     * 
     * @return the max diameter that this cell can have, e.g. the diameter that this cell has if no other cell is around.
     */
    double getMaxDiameter();

    /**
     * 
     * @return the max radius that this cell can have, e.g. the radius that this cell has if no other cell is around.
     */
    double getMaxRadius();
}

File: alchemist-incarnation-biochemistry/src/main/kotlin/it/unibo/alchemist/model/implementations/properties/CircularDeformableCell.kt
/*
 * Copyright (C) 2010-2022, Danilo Pianini and contributors
 * listed, for each module, in the respective subproject's build.gradle.kts file.
 *
 * This file is part of Alchemist, and is distributed under the terms of the
 * GNU General Public License, with a linking exception,
 * as described in the file LICENSE in the Alchemist distribution's top directory.
 */

package it.unibo.alchemist.model.implementations.properties

import it.unibo.alchemist.model.Environment
import it.unibo.alchemist.model.Node
import it.unibo.alchemist.model.implementations.molecules.Junction
import it.unibo.alchemist.model.implementations.positions.Euclidean2DPosition
import it.unibo.alchemist.model.interfaces.properties.CircularCellProperty
import it.unibo.alchemist.model.interfaces.properties.CircularDeformableCellProperty

/**
 * Base implementation of a [CircularCellProperty].
 */
class CircularDeformableCell @JvmOverloads constructor(
    environment: Environment<Double, Euclidean2DPosition>,
    override val node: Node<Double>,
    override val maximumDiameter: Double,
    override val rigidity: Double,
    override val junctions: MutableMap<Junction, MutableMap<Node<Double>, Int>> = LinkedHashMap(),
) : AbstractNodeProperty<Double>(node),
    CircularDeformableCellProperty,
    CircularCellProperty by CircularCell(
        environment,
        node,
        maximumDiameter * rigidity,
        junctions,
    ) {
    init {
        require(rigidity in 0.0..1.0) {
            "deformability must be between 0 and 1"
        }
    }

    override fun toString() = "${super.toString()}[maximumDiameter=$maximumDiameter, rigidity=$rigidity]"
}
